src/GoogleFitOnFhir/Services/GoogleFitDataImporter.cs

```diff
@@ -98,34 +96,67 @@ public async Task Import(string userId, CancellationToken cancellationToken)
-
-                // Add user id to payload
-                dataset.UserId = user.Id;
-
-                var jsonDataset = JsonConvert.SerializeObject(dataset);
-
-                _logger.LogInformation("Push Dataset: {0}", datasourceId);
+                datasetRequestTasks.Add(Task.Run(
+                    async () =>
+                    {
+                        string pageToken = null;
+
+                        // Make the Dataset requests, requesting the next page of data if necessary
+                        do
+                        {
+                            _logger.LogInformation("Query Dataset: {0}", datasourceId);
+                            var dataset = await _googleFitClient.DatasetRequest(
+                                tokensResponse.AccessToken,
+                                datasourceId,
+                                datasetId,
+                                cancellationToken,
+                                pageToken);
+
+                            // Save the NextPageToken
+                            pageToken = dataset.NextPageToken;
+
+                            // Add user id to payload
+                            dataset.UserId = user.Id;
+
+                            var jsonDataset = JsonConvert.SerializeObject(dataset);
+
+                            _logger.LogInformation("Push Dataset: {0}", datasourceId);
+
+                            // Create a batch of events for IoMT eventhub
+                            using var eventBatch = await _eventHubProducerClient.CreateBatchAsync(cancellationToken);
+
+                            // Push dataset to IoMT connector
+                            if (!eventBatch.TryAdd(new EventData(jsonDataset)))
+                            {
+                                throw new Exception("Event is too large for the batch and cannot be sent.");
+                            }
+
+                            // Use the producer client to send the batch of events to the event hub
+                            await _eventHubProducerClient.SendAsync(eventBatch, cancellationToken);
+                            _logger.LogInformation("A batch of events has been published for {0}.", datasourceId);
+                        }
+                        while (pageToken != null);
+                    }, cancellationToken));
+            }
 
-                // Push dataset to IoMT connector
-                if (!eventBatch.TryAdd(new EventData(jsonDataset)))
+            // Wait for the Dataset request tasks to finish
+            try
+            {
+                Task.WaitAll(datasetRequestTasks.ToArray(), cancellationToken);
+            }
+            catch (AggregateException ex)
+            {
+                foreach (var ie in ex.InnerExceptions)
                 {
-                    throw new Exception("Event is too large for the batch and cannot be sent.");
+                    _logger.LogError("{0}: {1}", ie.GetType().Name, ie.Message);
```

Comment:
LogError allows you to just pass the Exception. This is useful because it includes the stack trace and other exception data. You can also add an optional message.